crates/red_knot_python_semantic/resources/mdtest/dataclasses.md

```diff
@@ -197,18 +320,61 @@ class C:
 reveal_type(C.__init__)  # revealed: (*args: Any, **kwargs: Any) -> None
 ```
 
-### Dataclass with `init=False`
+### Dataclass with custom `__init__` method
 
-To do
+If a class already defines `__init__`, it is not replaced by the `dataclass` decorator.
 
-### Dataclass with custom `__init__` method
+```py
+from dataclasses import dataclass
 
-To do
+@dataclass(init=True)
+class C:
+    x: str
+
+    def __init__(self, x: int) -> None:
+        self.x = str(x)
+
+C(1)  # OK
+
+# TODO: should be an error
+C("a")
+```
+
+Similarly, if we set `init=False`, we still recognize the custom `__init__` method:
+
+```py
+@dataclass(init=False)
+class D:
+    def __init__(self, x: int) -> None:
+        self.x = str(x)
+
+D(1)  # OK
+D()  # error: [missing-argument]
+```
 
 ### Dataclass with `ClassVar`s
 
 To do
 
+### Return type of `dataclass(...)`
+
+A call like `dataclass(order=True)` returns a callable itself, which is then used as the decorator.
+We can store the callable in a variable and later use it as a decorator:
+
+```py
+from dataclasses import dataclass
+
+dataclass_with_order = dataclass(order=True)
+
+reveal_type(dataclass_with_order)  # revealed: <decorator produced by dataclasses.dataclass>
+
+@dataclass_with_order
+class C:
+    x: int
+
+C(1) < C(2)  # ok
+```
```

Comment:
@AlexWaygood I already added a couple of non-`order`-related tests in this PR, so I added the test that you suggested as well.